### PR TITLE
[ENH] Add lazy imports to speed up the time taken to load pyjanitor (part 2)

### LINKED ISSUE #1059: Enable lazy importing for external dependencies
I'd like to propose that we use lazy imports for our external dependencies.

Currently, importing pyjanitor results in about ~1-2 seconds of delay. While insignificant for an interactive data exploration setting, it becomes a significant latency burden if there are projects that depend on `pyjanitor`. 

There is a SPEC for lazy importing, which means this idea is getting community traction in the SciPy world: https://scientific-python.org/specs/spec-0001/

There is also a tool for making lazy loading easy: https://pypi.org/project/lazy_loader/, thereby giving us a path towards lazy loading.

Would love to get thoughts from the dev team!

### PR DESCRIPTION
**Pull request by asmirnov69**
# PR Description

**This PR resolves #1059**

Code changes are taken from original PR https://github.com/pyjanitor-devs/pyjanitor/pull/1060 and placed to current dev branch head.

# PR Checklist

Please ensure that you have done the following:

1. [X] PR in from a fork off your branch.
2. [X] If you're not on the contributors list, add yourself to `AUTHORS.md`.
3. [X] Add a line to `CHANGELOG.md` under the latest version header (i.e. the one that is "on deck") describing the contribution.

# Automatic checks

There will be automatic checks run on the PR. These include:

- Building a preview of the docs on Netlify
- Automatically linting the code
- Making sure the code is documented
- Making sure that all tests are passed
- Making sure that code coverage doesn't go down.

# Relevant Reviewers

@ericmjl

### CODE REVIEW

**Comment by ericmjl** [comment]
Hi @asmirnov69, thanks for taking the initiative on this PR!

Can I ask, when you benchmark import times on your own machine, do you observe a speedup in import times?

I'm asking b/c on a fresh Codespace, running the import benchmark gets me:

image

However, a subsequent run gets me:

image

As you probably can tell, the absolute first clean import on a machine will take ~6 seconds, and subsequent imports will be fast. I'm not sure whether this will be desired behaviour or not, though. What are your thoughts here?

**Comment by asmirnov69** [comment]
Hi @ericmjl 

Looks like first run is busy caching bytecode into `__pycache__` dirs - in your case it is way too slow. Difference and the timings you observe are much bigger that I can see on my ubuntu 22.04 4 cpu 16Gb in aws. All my tests are faster than 1 sec - including test with no cached bytecode. 
I will try to do one more test on fresh instance, will let you know if I will observe significant slow down on first time use - but I think something is happening in your system.

Few immediate observations:
- tuna.png doesn't go through pre-commit - file is ~800k which is bigger than allowed 500k
- simple timing shows improvement in import time ~30%. Removing of `__pycache__` dirs slow both tests down ~8%. The slowdown if -B cli options is used to prevent bytecode caching is the same 8% as expected
```
time python -c 'import janitor' # real time ~0.7 secs
time python -B -c 'import janitor' # 
python -X importtime -c 'import janitor' |& tail # look at the last line
python -B -X importtime -c 'import janitor' |& tail # 
```
30% improvement is smaller gain than you've seen (~50%). But looks like it make sense to use lazy_import in pyjanitor since gain is here and if anything goes wrong with future development it should be easy to change import statements.

<hr>
In general slow startup time of some script due to huge package dependencies could become the problem elsewhere too.
E.g. usual executables or shared objects sometimes loaded before program start using LD_PRELOAD (also used if you have circular dependencies). Maybe python3 team will add something like this in python itself in future.
<br/><br/>
But the real solution could be proper split of pyjanitor to pluggable modules (as mentioned in #826). pyjanitor list of mandatory dependencies is too long.

In long term I am for packing of pyjanitor using [optional dependencies](https://setuptools.pypa.io/en/latest/userguide/dependency_management.html#optional-dependencies) which I think is proper answer on how to implement the ask of #826: 

> Is the following possible?
> pip install pyjanitor[base]

BTW, I can explore what kind of visualization support can be added to my proposal #1176 to start addressing package dependencies problem. The idea could be to collect import dependencies into rdf triple form and then maybe have some useful diagram out of that - e.g. pyjanitor submodules dependencies graph.

**Comment by asmirnov69** [comment]
What should be done about codecov checks failures? Is it mandatory precondition to have all checks green to have contribution accepted?

**Comment by ericmjl** [comment]
@asmirnov69 thanks for the detailed response, and for putting in the time and effort to do profiling on your end.

Considering what you've written, I think we're good with the PR. 

With respect to codecov, the big ones to look out for are that all tests pass (primarily), and that code coverage on the *project* doesn't decrease significantly. In this case, a decrease by 0.41% is acceptable.

**Review by ericmjl — APPROVED** [review]
Approving the PR. Let's see how this works out on `dev` (i.e. make sure all tests still pass).

**Comment by asmirnov69** [comment]
@ericmjl do i need to do anything at this point? who is going to complete this PR with merge commit to dev?

**Comment by ericmjl** [comment]
We'll need one more review from the team. I just asked for @samukweku, @thatlittleboy, and @Zeroto521. As a rule of thumb, we try to go for 2 reviews (but time-bound to ~1-1.5 weeks of having a PR open.)

**Review by samukweku — APPROVED** [review]

File: AUTHORS.md
Comment (Zeroto521) [inline AUTHORS.md:112]: It could index all.
```suggestion
- [@asmirnov69](https://github.com/asmirnov69) | [contributions](https://github.com/pyjanitor-devs/pyjanitor/pulls?q=is%3Aclosed+mentions%3Aasmirnov69)
```

Comment (Zeroto521) [inline AUTHORS.md:112]: It could index all.
```suggestion
- [@asmirnov69](https://github.com/asmirnov69) | [contributions](https://github.com/pyjanitor-devs/pyjanitor/pulls?q=is%3Aclosed+mentions%3Aasmirnov69)
```

File: CHANGELOG.md
Comment (Zeroto521) [inline CHANGELOG.md:5]:
```suggestion
-   [ENH] Add lazy imports to speed up the time taken to load pyjanitor. Issue #1059 @asmirnov69.
```

File: janitor/__init__.py
Comment (Zeroto521) [inline janitor/__init__.py]: I'm unsure whether adding a suffix will have an impact on `bumpversion`.
Once the CI of `auto-release.yml` is trigged, the version will be like `0.24.0` or `0.24.0-lazy`?

File: janitor/functions/impute.py
Comment (Zeroto521) [inline janitor/functions/impute.py:7]: I saw all packages in `janitor/math.py` were wrapped by `lazy.load`.
But `janitor/functions/impute.py` only wrapped `scipy.stats`.

**Review by Zeroto521 — CHANGES_REQUESTED** [review]
Some small patches.

Since `lazy_loader` is imported in `janitor/__init__.py`. It's a necessary package.
So It need to add into `.requirements/base.in`.

File: janitor/__init__.py
Comment (asmirnov69) [reply janitor/__init__.py]: suffix -lazy to __version__ value  was introduced by me since I did't know what to do about version in the case 
of PR branch.
so feel free to drop it and set __version__ appropriately to current situation

File: janitor/functions/impute.py
Comment (asmirnov69) [reply janitor/functions/impute.py:7]: i didn't try to go beyond original @ericmjl code changes. 
it could be true that some newer additions on dev would require changes in imports. i assume we can address it later in subsequent PRs.

File: janitor/__init__.py
Comment (Zeroto521) [reply janitor/__init__.py]: I'd like to reverse this change to version first.

File: janitor/functions/impute.py
Comment (ericmjl) [reply janitor/functions/impute.py:7]: Good point, @Zeroto521. We can continue to improve coverage of lazy loading, so no worries here. Plus we might also benefit from Python having lazy loading by default! (I'm thinking of [PEP690](https://peps.python.org/pep-0690/) here.) Let's stick with what we have for now.

File: janitor/__init__.py
Comment (ericmjl) [reply janitor/__init__.py]: I agree with you, @Zeroto521, very astute catch that I didn't see before! @asmirnov69 could you remove the `-lazy` here? We'll take care of versioning in our build pipeline. That is probably the only thing I'll ask of before we merge.

Comment (asmirnov69) [reply janitor/__init__.py]: will do:
- \_\_version\_\_ will be 0.23.1
- lazy_loader will be added to .requirements/base.in

Comment (asmirnov69) [reply janitor/__init__.py]: done

**Review by Zeroto521 — APPROVED** [review]

**Comment by ericmjl** [comment]
Three approvers! I am going to hit the merge button :smile:.

**Pull request merged**
